### DISCUSSION TITLE
Scaling to 0 with a CPU scaler

### DISCUSSION THREAD

**Discussion by aryan9600** [kedacore/keda#2736] [Q&A / Need Help]
Hello, I'd like to ask is it at all possible to scale the target deployment to 0 replicas when using a CPU scaler? From https://keda.sh/docs/2.6/scalers/cpu/
> This scaler will never scale to 0 and even when user defines multiple scaler types (eg. Kafka + cpu/memory, or Prometheus + cpu/memory), the deployment will never scale to 0.

Is this true, even when `spec.IdleReplicaCount` is set to 0? If so, could I please understand the reason behind this functionality?

**Comment by tomkerkhove** [comment]
Scale to 0 for CPU or memory is not required because if it goes to 0, nothing is running.

If nothing is running, CPU nor memory would ever go up because there is nothing running anymore.

Makes sense?

**Comment by aryan9600** [comment]
Yeah, but a use case might arise where there are other controllers that can scale the deployment from 0 to 1 and vice versa. KEDA can scale the deployment between the specified min and max replica count based on the cpu consumption only when there's a minimum of one pod running. The HPA behaves in a similar manner, if I am not mistaken.

**Reply by tomkerkhove** [reply]
KEDA assumes that it's the only autoscaler so that scenario would never happen, unless you would use multiple scalers or what are you trying to achieve?

**Reply by aryan9600** [reply]
Some controllers need to scale a deployment down to 0, in order to make the deployment inactive and then scale it back to 1 to make the deployment active again. This could be independent of the actual scaling of replicas that is required when the deployment is active.

**Reply by tomkerkhove** [reply]
I'm not sure I'm getting the scenario - Sorry.

So you want to scale the workload with KEDA but have another controller scale it as well; is that it?

**Reply by aryan9600** [reply]
Kind of yes. KEDA would scale the the deployment only when there are 1 or more replicas, but it should leave the deployment alone if it  has been scaled down to zero replicas. Maybe we could have an annotation that enables this, or maybe the CPU and memory scalers could infer that if `idleReplicaCount` is set to 0, then it shouldn't scale the deployment back to the `minReplicaCount`?

**Reply by tomkerkhove** [reply]
You might want to check https://github.com/kedacore/keda/issues/944 but we do not encourage people to mix autoscalers. We'd probably actually discourage you to do it.

May I ask why you'd want to use another autoscaler? If so, maybe https://github.com/kedacore/keda/issues/944 is what you need where that other controller just tells it to stop autoscaling and what replica count wants.

**Reply by aryan9600** [reply]
Yep, this looks apt. We want to add KEDA support for [Flagger](https://flagger.app). Flagger scales the canary deployment to 1 during a canary analysis (which can be further autoscaled using a HPA) and then it scales the deployment down to 0 when the analysis is over. With KEDA, the deployment is scaled back up to the `minReplicaCount` when Flagger scales the deployment down to 0, which is unwanted.

**Reply by tomkerkhove** [reply]
Then I think using the above is ideal for you.

I'm not sure about Flagger's details, but if it has a CRD we can scale that resources as a scaleTarget as well.

**Comment by asobti** [comment]
For anyone who comes across this issue, this has been made possible since Keda v2.11
https://github.com/kedacore/keda/pull/4346
